lib/xmlrpc/server.rb

```diff
@@ -532,6 +532,49 @@ def http_write(body, status, header)
 end
 
 
+# Implements a XML-RPC server, which works with Rack
+class RackServer < BasicServer
+
+  # This method processes a XML-RPC method call and sends the answer
+  # back to the client.
+  def call(env)
+    length = env['CONTENT_LENGTH'].to_i
+
+    return http_error(405, "Method Not Allowed") unless env['REQUEST_METHOD'] == "POST"
+    return http_error(400, "Bad Request")        unless parse_content_type(env['CONTENT_TYPE']).first == "text/xml"
+    return http_error(411, "Length Required")    unless length > 0
+
+    req = Rack::Request.new(env)
+    data = req.body.read(length)
+
+    return http_error(400, "Bad Request")        if data.nil? or data.bytesize != length
+
+    [200, { "Content-type" => "text/xml; charset=utf-8" }, [process(data)]]
+  end
+
+
+  private
+
+  def http_error(status, message)
+    err = "#{status} #{message}"
+    msg = <<-"MSGEND"
+      <html>
+        <head>
+          <title>#{err}</title>
+        </head>
+        <body>
+          <h1>#{err}</h1>
+          <p>Unexpected error occurred while processing XML-RPC request!</p>
+        </body>
+      </html>
+    MSGEND
```

Comment:
```suggestion
    msg = <<-HTML
      <html>
        <head>
          <title>#{err}</title>
        </head>
        <body>
          <h1>#{err}</h1>
          <p>Unexpected error occurred while processing XML-RPC request!</p>
        </body>
      </html>
    HTML
```

Reply:
We've got this exact same code snippet two more times in the existing code (`CGIServer` and `ModRubyServer`). I would prefer to keep this the same for now, and as a future work extract it to some shared code and remove this duplication. It's easier to find all the occurrences if we keep this exactly the same for now.
Once the code has been deduplicated, this improvement can be applied.